File: BinarySearch/2528.Maximize-the-Minimum-Powered-City/2528.Maximize-the-Minimum-Powered-City.py
class Solution:
    def maxPower(self, stations: List[int], r: int, k: int) -> int:
        n = len(stations)
        left, right = 0, 10**18
        while left < right:
            mid = right - (right - left) // 2

            if self.isOK(stations, r, k, mid):
                left = mid
            else:
                right = mid -1
        return left
    
    def isOK(self, stations: List[int], r: int, k: int, mid: int) -> bool:
        n = len(stations)
        sum = 0
        arr = stations[:]

        for i in range(min(r, n)):
            sum += arr[i]

        for i in range(n):
            if i+r < n:
                sum += arr[i+r]
            if i-r-1 >= 0:
                sum -= arr[i-r-1]

            if sum >= mid:
                continue
            d = mid - sum
            if k < d: 
                return False
            pos = min(n - 1, i + r)
            arr[pos] += d
            sum = mid
            k -= d
        
        return True

File: BinarySearch/2528.Maximize-the-Minimum-Powered-City/2528.Maximize-the-Minimum-Powered-City.java
class Solution {
    public long maxPower(int[] stations, int r, int k) {
        int n = stations.length;
        long left = 0, right = Long.MAX_VALUE;
        while (left < right) {
            long mid = right - (right - left) / 2;
            long[] temp = new long[n];
            for (int i = 0; i < n; i++) {
                temp[i] = (long)stations[i];
            }

            if (isOK(temp, r, (long)k, mid)) {
                left = mid;
            } else {
                right = mid -1;
            }
        }
        return left;
    }

    private boolean isOK(long[] stations, int r, long k, long mid) {
        int n = stations.length;
        long sum = 0;
        for (int i = 0; i <= r-1; i++) sum += stations[i];

        for (int i = 0; i < n; i++) {
            // slide the window to right 1 postion and out left 1 position
            if (i+r < n) 
                sum += stations[i+r];
            if (i-r-1 >= 0) sum -= stations[i-r-1];

            if (sum >= mid) continue;
            
                long d = mid - sum;
                if (k < d) return false;
                stations[Math.min(n-1, i+r)] += d;
                k -= d;
                sum += d;
            

        }
        return true;
    }
}
// ---- x --- r distance
// k, 
// for every range will reach out binary search current value
// not match to add k into the most right postion on this window subarray
